src/backend/distributed/shardsplit/shardsplit_logical_replication.c

```diff
@@ -299,483 +176,92 @@ CreateTargetNodeConnectionsForShardSplit(List *shardSplitSubscribersMetadataList
+	}
 
-		shardSplitSubscriptionMetadataList = lappend(shardSplitSubscriptionMetadataList,
-													 shardSplitSubscriberMetadata);
+	List *shardIntervalList = NIL;
+	foreach_ptr(shardIntervalList, shardGroupSplitIntervalListList)
+	{
+		ShardInterval *shardInterval = NULL;
+		WorkerNode *workerPlacementNode = NULL;
+		forboth_ptr(shardInterval, shardIntervalList, workerPlacementNode,
+					workersForPlacementList)
+		{
+			NodeAndOwner key;
+			key.nodeId = workerPlacementNode->nodeId;
+			key.tableOwnerId = TableOwnerOid(shardInterval->relationId);
+
+			bool found = false;
+			publication = (PublicationInfo *) hash_search(
+				ShardInfoHashMapForPublications,
+				&key,
+				HASH_FIND,
+				&found);
+			if (!found)
+			{
+				ereport(ERROR, errmsg("Could not find publication matching a split"));
+			}
+			publication->subscription->newShards = lappend(
+				publication->subscription->newShards, shardInterval);
+		}
 	}
 
-	return shardSplitSubscriptionMetadataList;
+	return subscriptionInfoList;
 }
 
 
 /*
- * Creates a 'ShardSplitSubscriberMetadata' structure for given table owner, node id.
+ * Creates a 'SubscriptionInfo' structure for given table owner, node id.
  * It scans the list of 'ReplicationSlotInfo' to identify the corresponding slot
  * to be used for given tableOwnerId and nodeId.
  */
-ShardSplitSubscriberMetadata *
-CreateShardSplitSubscriberMetadata(Oid tableOwnerId, uint32 nodeId,
-								   List *replicationSlotInfoList)
+SubscriptionInfo *
+CreateSubscriptionInfo(Oid tableOwnerId, uint32 nodeId,
+					   List *replicationSlotInfoList)
 {
-	ShardSplitSubscriberMetadata *shardSplitSubscriberMetadata = palloc0(
-		sizeof(ShardSplitSubscriberMetadata));
-	shardSplitSubscriberMetadata->tableOwnerId = tableOwnerId;
+	SubscriptionInfo *subscription = palloc0(
+		sizeof(SubscriptionInfo));
+	subscription->name = SubscriptionName(SHARD_SPLIT, tableOwnerId);
+	subscription->tableOwnerId = tableOwnerId;
+	subscription->temporaryOwnerName =
+		SubscriptionRoleName(SHARD_SPLIT, tableOwnerId);
+	subscription->targetConnection = NULL;
 
 	/*
 	 * Each 'ReplicationSlotInfo' belongs to a unique combination of node id and owner.
 	 * Traverse the slot list to identify the corresponding slot for given
 	 * table owner and node.
 	 */
-	char *tableOwnerName = GetUserNameFromId(tableOwnerId, false);
-	ReplicationSlotInfo *replicationSlotInfo = NULL;
-	foreach_ptr(replicationSlotInfo, replicationSlotInfoList)
+	ReplicationSlotInfo *replicationSlot = NULL;
+	foreach_ptr(replicationSlot, replicationSlotInfoList)
 	{
-		if (nodeId == replicationSlotInfo->targetNodeId &&
-			strcmp(tableOwnerName, replicationSlotInfo->tableOwnerName) == 0)
+		if (nodeId == replicationSlot->targetNodeId &&
+			tableOwnerId == replicationSlot->tableOwnerId)
 		{
-			shardSplitSubscriberMetadata->slotInfo = replicationSlotInfo;
+			subscription->replicationSlot = replicationSlot;
+
 			break;
 		}
 	}
```

Comment:
Would be nice to assert on `subscription->replicationSlot != NULL` and throw an error before return.